src/py_zfs_vdev.c

```diff
@@ -0,0 +1,320 @@
+	return (0);
+}
+
+static
+void py_zfs_vdev_dealloc(py_zfs_vdev_t *self) {
+	Py_CLEAR(self->pylibzfsp);
+	Py_CLEAR(self->parent);
+	Py_CLEAR(self->path);
+	Py_CLEAR(self->type);
+	Py_TYPE(self)->tp_free((PyObject *)self);
+}
+
+static
+PyObject *py_zfs_vdev_asdict(PyObject *self, PyObject *args) {
+	Py_RETURN_NONE;
+}
+
+static
+PyObject *py_zfs_vdev_get_type(py_zfs_vdev_t *self, void *extra) {
+	if (self == NULL || self->type == NULL)
+		Py_RETURN_NONE;
+	return (Py_NewRef(self->type));
+}
+
+static
+PyObject *py_zfs_vdev_get_path(py_zfs_vdev_t *self, void *extra) {
+	if (self == NULL || self->path == NULL)
+		Py_RETURN_NONE;
+	return (Py_NewRef(self->path));
+}
+
+static
+PyObject *py_zfs_vdev_get_name(PyObject *self, PyObject *args) {
+	py_zfs_vdev_t *v = (py_zfs_vdev_t *)self;
+	PyObject *out = NULL;
+	char *name = NULL;
+
+	Py_BEGIN_ALLOW_THREADS
+	PY_ZFS_LOCK(v->pylibzfsp);
+	name = zpool_vdev_name(v->pylibzfsp->lzh, v->zhp, v->vdev_tree,
+	    VDEV_NAME_TYPE_ID);
+	PY_ZFS_UNLOCK(v->pylibzfsp);
+	Py_END_ALLOW_THREADS
+
+	out = PyUnicode_FromString(name);
+	free(name);
+	return (out);
+}
+
+static
+PyObject *py_zfs_vdev_get_guid(PyObject *self, PyObject *args) {
+	py_zfs_vdev_t *v = (py_zfs_vdev_t *)self;
+	uint64_t guid;
+
+	Py_BEGIN_ALLOW_THREADS
+	guid = fnvlist_lookup_uint64(v->vdev_tree, ZPOOL_CONFIG_GUID);
+	Py_END_ALLOW_THREADS
+
+	return (PyLong_FromUnsignedLongLong(guid));
+}
+
+static
+const char *vdev_get_status_impl(nvlist_t *tree) {
+	vdev_stat_t *vs;
+	uint_t c;
+	const char *state = NULL;
+	Py_BEGIN_ALLOW_THREADS
+	if (nvlist_lookup_uint64_array(tree, ZPOOL_CONFIG_VDEV_STATS,
+	    (uint64_t **)&vs, &c) == 0) {
+		state = zpool_state_to_name(vs->vs_state, vs->vs_aux);
+	}
+	Py_END_ALLOW_THREADS
+	return (state);
+}
+
+static
+PyObject *py_zfs_vdev_get_status(PyObject *self, PyObject *args) {
+	py_zfs_vdev_t *v = (py_zfs_vdev_t *)self;
+	const char *state = vdev_get_status_impl(v->vdev_tree);
+	if (state == NULL)
+		Py_RETURN_NONE;
+	return (PyUnicode_FromString(state));
+}
+
+static
+PyObject *py_zfs_vdev_get_size(PyObject *self, PyObject *args) {
+	py_zfs_vdev_t *v = (py_zfs_vdev_t *)self;
+	uint64_t asize = 0;
+	uint64_t ashift = 0;
+	uint64_t size;
+	boolean_t found;
+
+	Py_BEGIN_ALLOW_THREADS
+	found = (nvlist_lookup_uint64(v->vdev_tree, ZPOOL_CONFIG_ASIZE,
+	    &asize) == 0 && nvlist_lookup_uint64(v->vdev_tree,
+	    ZPOOL_CONFIG_ASHIFT, &ashift) == 0);
+	Py_END_ALLOW_THREADS
+
+	if (!found)
+		Py_RETURN_NONE;
+	size = asize << ashift;
+	return (PyLong_FromUnsignedLongLong(size));
+}
+
+static
+PyObject *py_zfs_vdev_get_children(PyObject *self, PyObject *args) {
+	py_zfs_vdev_t *v = (py_zfs_vdev_t *)self;
+	nvlist_t  **child;
+	uint_t children;
+	boolean_t found;
+	PyObject *list, *obj;
+
+	Py_BEGIN_ALLOW_THREADS
+	found = (nvlist_lookup_nvlist_array(v->vdev_tree, ZPOOL_CONFIG_CHILDREN,
+	    &child, &children) == 0);
+	Py_END_ALLOW_THREADS
+
+	if (!found)
+		Py_RETURN_NONE;
+
+	list = PyList_New(0);
```

Comment:
If you know that the there won't be toc/tou issues, then you can pre-allocate the list to the required size, then use `PyList_SET_ITEM`.

Reply:
On second thought. It'd probably be better to do something like:
```
PyTuple_New(children);
```

Then `PyTuple_SET_ITEM(obj);` in your loop. Tuple is more nature here since it's immutable (users can't append / delete items from it in expectation that it changes vdev configuration).